File: Problems & Solutions/1626. Best Team With No Conflicts/31 January 2023/Approach 2 - Bottom-Up Dynamic Programming/Best Team With No Conflicts.cpp
class Solution {
public:
    int findMaxScore(vector<pair<int, int>>& ageScorePair) {
        int n = (int) ageScorePair.size();
        int answer = 0;
        
        vector<int> dp(n);
        for (int i = 0; i < n; i++) {
            dp[i] = ageScorePair[i].second;
            answer = max(answer, dp[i]);
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = i - 1; j >= 0; j--) {
                if (ageScorePair[i].second >= ageScorePair[j].second) {
                    dp[i] = max(dp[i], ageScorePair[i].second + dp[j]);
                }
            }
            answer = max(answer, dp[i]);
        }
        
        return answer;
    }
    
    int bestTeamScore(vector<int>& scores, vector<int>& ages) {
        vector<pair<int, int>> ageScorePair;
        for (int i = 0; i < scores.size(); i++) {
            ageScorePair.push_back({ages[i], scores[i]});
        }
        
        sort(ageScorePair.begin(), ageScorePair.end());
        return findMaxScore(ageScorePair);
    }
};

File: Problems & Solutions/1626. Best Team With No Conflicts/31 January 2023/Approach 2 - Bottom-Up Dynamic Programming/Best Team With No Conflicts.java
class Solution {
    private int findMaxScore(int[][] ageScorePair) {
        int n = ageScorePair.length;
        int answer = 0;

        int[] dp = new int[n];
        for (int i = 0; i < n; i++) {
            dp[i] = ageScorePair[i][1];
            answer = Math.max(answer, dp[i]);
        }


        for (int i = 0; i < n; i++) {
            for (int j = i - 1; j >= 0; j--) {
                if (ageScorePair[i][1] >= ageScorePair[j][1]) {
                    dp[i] = Math.max(dp[i], ageScorePair[i][1] + dp[j]);
                }
            }
            answer = Math.max(answer, dp[i]);
        }

        return answer;
    }

    public int bestTeamScore(int[] scores, int[] ages) {
        int N = ages.length;
        int[][] ageScorePair = new int[N][2];

        for (int i = 0; i < N; i++) {
            ageScorePair[i][0] = ages[i];
            ageScorePair[i][1] = scores[i];
        }

        Arrays.sort(ageScorePair, (a,b) -> a[0] == b[0] ? a[1]-b[1] : a[0]-b[0]);
        return findMaxScore(ageScorePair);
    }
}